### PR TITLE
feat(eslint-plugin): Add unified-signature rule

### PR DESCRIPTION
**Pull request by armanio123**
Adding tslint's rule `unified-signature`.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/unified-signatures.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+```
+```ts
+f(): void;
+f(...x: number[]): void;
+```
+
+Examples of **correct** code for this rule:
+
+```ts
+function f(x: number | stting): void;
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/unified-signatures.md]:
```suggestion
function f(x: number | string): void;
```

Comment (JamesHenry) [reply packages/eslint-plugin/docs/rules/unified-signatures.md]: image

Comment (JamesHenry) [reply packages/eslint-plugin/docs/rules/unified-signatures.md]:
```ts
interface Sting extends EnglishmanInNewYork {}
```

Comment (JamesHenry) [reply packages/eslint-plugin/docs/rules/unified-signatures.md]: Ok, I'll stop...

Comment (armanio123) [reply packages/eslint-plugin/docs/rules/unified-signatures.md]: Haha good one, thanks for the laugh. Fixed.

File: packages/eslint-plugin/ROADMAP.md [hunk 2]
```diff
@@ -1,10 +1,10 @@
 ﻿# Roadmap
 
-✅ (28) = done  
-🌟 (79) = in ESLint core  
-🔌 (33) = in another plugin  
-🌓 (16) = implementations differ or ESLint version is missing functionality  
-🛑 (70) = unimplemented
+✅ (29) = done
```
Comment (armano2) [inline packages/eslint-plugin/ROADMAP.md]: an you restore white-spaces?
https://github.com/typescript-eslint/typescript-eslint/blob/c89abbdea5298fccebca4e66fc362037a2ea1fec/packages/eslint-plugin/ROADMAP.md

i think we should convert them to `<br>`, and same below

File: packages/eslint-plugin/lib/rules/unified-signatures.js [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+                        });
+                        break;
+                    }
+                    case "extra-parameter": {
+                        const { extraParameter, otherSignature } = unify;
+                        const lineOfOtherOverload = only2 ? undefined : otherSignature.loc.start.line;
+
+                        context.report({
+                            loc: extraParameter.loc,
+                            message: extraParameter.type === Syntax.RestElement
```
Comment (armano2) [inline packages/eslint-plugin/lib/rules/unified-signatures.js]: can you convert `message` to `messageId` with `data`
you can find example here: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/lib/rules/no-type-alias.js#L32

https://eslint.org/docs/developer-guide/working-with-rules#messageids

File: packages/eslint-plugin/lib/rules/unified-signatures.js [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+     * @param {number} [otherLine]
+     * @returns {string}
+     */
+    function FAILURE_STRING_START(otherLine) {
+      // For only 2 overloads we don't need to specify which is the other one.
+      const overloads =
+        otherLine === undefined
+          ? 'These overloads'
+          : `This overload and the one on line ${otherLine}`;
+      return `${overloads} can be combined into one signature`;
```
Comment (armano2) [inline packages/eslint-plugin/lib/rules/unified-signatures.js]: messages should be migrated to `messageId`, you can find usages in other rules or in documentation:

https://eslint.org/docs/developer-guide/working-with-rules#messageids

**Comment by JamesHenry** [comment]
Apologies for the conflicts, we decided to remove the counts from the ROADMAP here: https://github.com/typescript-eslint/typescript-eslint/pull/225

They are too awkward to maintain manually, so this is the last time you will have to deal with syncing that up!

**Comment by JamesHenry** [comment]
@armanio123 Thanks again for your contribution, when do you think you'll be able to pick this up again?

File: packages/eslint-plugin/lib/rules/unified-signatures.js [hunk 5]
```diff
@@ -0,0 +1,616 @@
+/**
+ * @fileoverview Warns for any two overloads that could be unified into one by using a union or an optional/rest parameter.
+ * @author Armando Aguirre
```
Comment (JamesHenry) [inline packages/eslint-plugin/lib/rules/unified-signatures.js]: Sorry we've changed to using all-contributors now and removed all other occurrences of these file-level author comments, please could you remove this

File: packages/eslint-plugin/lib/rules/unified-signatures.js (hunk 3 above)
Comment (armanio123) [reply packages/eslint-plugin/lib/rules/unified-signatures.js]: Converted.

File: packages/eslint-plugin/lib/rules/unified-signatures.js (hunk 4 above)
Comment (armanio123) [reply packages/eslint-plugin/lib/rules/unified-signatures.js]: Done.

File: packages/eslint-plugin/lib/rules/unified-signatures.js (hunk 5 above)
Comment (armanio123) [reply packages/eslint-plugin/lib/rules/unified-signatures.js]: Removed.

**Comment by armanio123** [comment]
Added the changes requested, migrated to TS and addressed the comments as well. Let me know if there's anything else.

File: packages/eslint-plugin/ROADMAP.md [hunk 6]
```diff
@@ -34,7 +34,7 @@
 | [`promise-function-async`]        | ✅  | [`@typescript-eslint/promise-function-async`]        |
 | [`typedef`]                       | 🛑  | N/A                                                  |
 | [`typedef-whitespace`]            | ✅  | [`@typescript-eslint/type-annotation-spacing`]       |
-| [`unified-signatures`]            | 🛑  | N/A                                                  |
+| [`unified-signatures`]            | ✅  | [`@typescript-eslint/unified-signatures`]            |
```
Comment (bradzacher) [inline packages/eslint-plugin/ROADMAP.md]: need to add the link at the bottom of the file or else this won't work

Comment (armanio123) [reply packages/eslint-plugin/ROADMAP.md]: Fixed.

**Comment by JamesHenry** [comment]
Thanks for working through all the feedback, @armanio123!

**Comment by JamesHenry** [comment]
@uniqueiniquity is going to apply this to the TS repo this week :)

**Pull request merged**